### DISCUSSION TITLE
v3.0.0-rc.1

### DISCUSSION THREAD

**Discussion by rexxars** [sanity-io/sanity#3843] [Announcements]
# Features

- Workspaces can now be given a custom `subtitle`, and will not use the dataset name by default
- TypeScript is now the default option for new projects (you are still prompted to confirm - answering no will still give you a javascript project)

# 🐛 Notable bugfixes
- Fixes incorrect command in help text after initializing a new project
- Fixes typing of `defineType` not allowing specified `components`
- Fixes incorrect version being printed from `sanity --version`

# 📓 Full changelog
Author | Message | Commit
------------ | ------------- | -------------
Espen Hovlandsdal | fix(cli): use correct command (`dev`, not `start`) in post-init text | 310e89b806
Espen Hovlandsdal | feat(cli): use typescript by default for new projects | d0806be7a9
Marius Lundgård | chore(deps): upgrade `@sanity/pkg-utils` | c2978cd46b
Espen Hovlandsdal | fix(cli): read version in runtime instead of bundling | d5f5afca02
Herman Wikner | fix(core): broken error screen ui | fe5a6bcafe
Herman Wikner | dev(test-studio): update `sanity.config` | 48968b6e27
Herman Wikner | dev(core): add `WorkspacePreview` workshop story | 0f36a5c0f1
Herman Wikner | feat(core): make workspace `subtitle` configurable | 2c60fe9ebc

<hr /><em>This discussion was created from the release <a href='https://github.com/sanity-io/sanity/releases/tag/v3.0.0-rc.1'>v3.0.0-rc.1</a>.</em>

**Comment by alex289** [comment]
This release breaks my next app. 
Error while building my next app on Vercel/locally:
```sh
./node_modules/.pnpm/sanity@3.0.0-rc.1_maey765qzsqrp6bbazpxj7endy/node_modules/sanity/lib/exports/index.d.ts:9807:10
Type error: Property or signature expected.

  9805 | declare module '@sanity/types' {
  9806 |     interface ArrayDefinition {
> 9807 |          */
       |          ^
  9808 |         components?: ArrayOfObjectsComponents | ArrayOfPrimitivesComponents;
  9809 |     }
  9810 |     interface BlockDefinition {
```

My Repo: https://github.com/alex289/Portfolio

**Reply by alex289** [reply]
Fixed in 3.0.0-rc.2

**Comment by bryandowning** [comment]
> Fixes typing of `defineType` not allowing specified components

This did indeed fix the TS error for `components` fields within `defineType` -> `defineField`. 🎉 

...but not for `defineField` functions that exist outside a `defineType` function:

```
export const color = fieldGenerator(
  defineField({
    type: 'color',
    name: 'color' as string,
    title: 'Color',
    // TODO: remove after sanity fixes this type error
    // @ts-ignore
    components: {
      input: ColorPicker,
    },
  })
)
```

(tested in rc.2)

**Reply by snorrees** [reply]
I am unable to reproduce this directly unfortunatly. This compile for me on rc.2:

```ts
const aa = defineField({
  type: "color",
  name: "color",
  components: {
    input: (inputProps) => null,
  },
});

```

However, the following gives a compilation error on the `input` line:

```ts
const aa = defineField({
  type: "color",
  name: "color",
  components: {
    // type the shape of the object props value
    // gives compilation error here
    input: (a: ObjectInputProps<{ someField: string }>) => null,
  },
});
```

I suspect there is a case to be made that we should have the default value type be `Record<string, any>` and not `Record<string, unknown>` to allow type-narrowing, and a bit less pain in the general case. In fact, I think we should go right ahead and change this.

What is the prop-types of the `ColorPicker` component @bryandowning?
Could you try using `ObjectInputProps` directly, and cast to your "true" value type within the component?

Ie, something like
```ts
function ColorPicker(props: ObjectInputProps) {
    const value = props.value as MyColorType | undefined
   //...
}
```

I realize you have your `@ts-ignore` on the `components` line and is seeing another error, but I'm just trying to figure out what is going on.

If you hoist the `defineField` out of `fieldGenerator`, does that change behaviour?

Oh, and are you using the `@sanity/color-input` plugin or not? (It adds it own "IntrinsicType" for color, which could possibly change things).

**Reply by bryandowning** [reply]
Thanks for the detailed response @snorrees!

Very strange, as I'm not seeing the same behavior as you (I even restarted my TS server in VS Code).

Given the following:
```js
export const objectField = defineField({
  type: "color",
  name: "color",
  components: {
    input: (inputProps) => null,
  },
});

export const typeWithField = defineType({
  name: 'sometype',
  title: 'Some Type',
  type: 'document',
  fields: [
    defineField({
      type: "color",
      name: "color",
      components: {
        input: (inputProps) => null,
      },
    }),
    defineField({
      type: "string",
      name: "color",
      components: {
        input: (inputProps) => null,
      },
    })
  ]
})
```

I get the following TS error on `objectField` and on the `type: 'color'` field on `typeWithField`:

```
Argument of type '{ type: "color"; name: "color"; components: { input: (inputProps: any) => null; }; }' is not assignable to parameter of type '{ type: "color"; name: "color"; } & BaseSchemaDefinition & { type: "color"; options?: unknown; validation?: SchemaValidationValue; initialValue?: any; preview?: PreviewConfig<...> | undefined; } & { ...; } & FieldDefinitionBase'.
  Object literal may only specify known properties, and 'components' does not exist in type '{ type: "color"; name: "color"; } & BaseSchemaDefinition & { type: "color"; options?: unknown; validation?: SchemaValidationValue; initialValue?: any; preview?: PreviewConfig<...> | undefined; } & { ...; } & FieldDefinitionBase'.ts(2345)
```

![Screen Shot 2022-11-08 at 1 45 54 PM](https://user-images.githubusercontent.com/468219/200682565-26d76043-ee4f-4f85-853c-bc22e405b3b9.png)


...elsewhere I've got this going on (which I'm assuming works because the field type is `array`):
![Screen Shot 2022-11-08 at 1 48 03 PM](https://user-images.githubusercontent.com/468219/200683027-69ba138b-afd4-4b86-8fca-e4bc2002ed29.png)


RE: the rest of your questions:

- I was actually already using `ObjectInputProps`, but now I'm totally gonna cast the whole `value` prop like you did instead if this ickyness:
```js
export function ColorPicker(props: ObjectInputProps) {
  const {onChange, value} = props
  const valueRgba = value?.rgba as string
```
- I rolled my own color picker input/object — no use of @sanity/color-input
- As shown above, using `fieldGenerator` doesn't have any impact on the TS error I'm seeing...but here's the code for it in case it helps.
```js
/**
 * This function returns a funtion that allows the caller to merge additional
 * schema fields onto a shared field.
 *
 * Example usage:
 *
 * export const title = fieldGenerator(
 *   defineField({
 *     type: 'string',
 *     name: 'title',
 *     title: 'Title',
 *   })
 * )
 *
 * title({fieldset: 'My Fieldset'})
 */
export function fieldGenerator<T>(base: T): (schemaFields?: Partial<T>) => T {
  return (schemaFields) => ({...base, ...schemaFields})
}
```

**Reply by snorrees** [reply]
## The problem
Just reproduced in a clean studio. Not having @sanity/color-input, was key. Sorry for not checking sooner 😓 

@sanity/color-input, which as I alluded to, adds an "intrinsic" `color`, type, making it resolve to basically ObjectDefinition. It stopped me from seeing this.

## What we should fix

Testing in clean studio makes `color` resolve to the generic fallback type `TypeAliasDefinition`. 
We did not add components to `TypeAliasDefinition`  when we narrowed props for all the intrinsic types. This is an oversight, so lets fix in the next release.

I was going to show you how you can "monkey-patch" `TypeAliasDefinition` until then, but now I realize that it is a `type` and not an `interface`. The latter is extendable, the former is not, unfortunately. We should fix this also.
 
 ## What you can do right now
 
 Now, let me instead show you how you can add color as an intrinsic type yourself, if you want. It will fix your problem in lieu of a new rc version. I also personally really like this approach, as you can extend the standard type schema with your own stuff:
 
 ```ts
 // anywhere in your studio, add this
 import { ObjectDefinition } from 'sanity'

// define your color schema, based on object-schema
export interface ColorDefinition extends Omit<ObjectDefinition, 'type' | 'fields'> {
  type: 'color'
}

/* example if you wanted to add custom option support
export interface ColorDefinition extends Omit<ObjectDefinition, 'type' | 'fields' | 'options'> {
  type: 'color'
  options: ColorOptions
  //or go real specific and define components and spesific object props
}
*/

declare module 'sanity' {
  // makes type: 'color' narrow correctly when using defineTyp/defineField/defineArrayMember
  export interface IntrinsicDefinitions {
    color: ColorDefinition
  }
}

//now, anywhere else, it should work

export const objectField = defineField({
  type: "color",   // color resolves to an intrinsic type now, so you get typesafety according to your own definition
  name: "color",
  //components exist on ObjectDefinition unlike TypeAliasDefinition, so you are good now :)
  components: {
    input: (inputProps) => null,
  },
});
 ```

**Reply by bryandowning** [reply]
You're a hero — thanks a lot for all this great info!
